File: ArrayProblems/src/my/com/app/ShortestWordDistance.py
from collections import defaultdict

class Solution(object):
    def shortestDistance(self, wordsDict, word1, word2):
        # this function finds the shortest distance between two words in a list of words.
        # another approach is to use a dictionary to store the indices of each word and then find the minimum distance between the two words.   
        # space complexity is O(n) and time complexity is O(n) where n is the length of the wordsDict.
        """
        :type wordsDict: List[str]
        :type word1: str
        :type word2: str
        :rtype: int
        
        indxDict = defaultdict(list)
        for i, s in enumerate(wordsDict):
            indxDict[s].append(i)
            
        l1 = indxDict[word1]
        l2 = indxDict[word2]

        i ,j = 0 , 0
        minDist = len(wordsDict)
        while i<len(l1) and j<len(l2):
            minDist = min( minDist , abs( l1[i] - l2[j]))
            if l1[i] < l2[j]:
                i+=1
            else:
                j+=1
        """

#         Given a list of words and two words, find the shortest distance between the two words in the list.
# this approachs useds constant space and O(n) time complexity
        p1, p2 = len(wordsDict), len(wordsDict)
        minDist = len(wordsDict)
        for i, s in enumerate(wordsDict):
            if s == word1:
                p1 = i
                minDist = min( minDist , abs( p1 - p2))
            elif s == word2:
                p2 = i
                minDist = min( minDist , abs( p1 - p2))
                
        return minDist
        

File: CustomDataStructures/src/my/com/app/ShortestWordDistance.java
package src.my.com.app;

import java.util.*;

// approach :
// 1. create a map of words and their indices
// 2. for each query, retrieve the indices of the two words
// 3. use two pointers to find the minimum distance between the indices of the two words
// 4. return the minimum distance found
// Time Complexity: O(n + m) where n is the number of words in the dictionary
// and m is the number of queries
// Space Complexity: O(n) for storing the indices of the words in the map

class WordDistance {

    HashMap<String , List<Integer>> idxMap;

    public WordDistance(String[] wordsDict) {
        idxMap = new HashMap();

        for (int i=0 ; i<wordsDict.length; i++){
            idxMap.computeIfAbsent(wordsDict[i], (v) ->  new ArrayList()).add(i);
        }    
    }
    
    public int shortest(String word1, String word2) {
        List<Integer> l1 = idxMap.get(word1);
        List<Integer> l2 = idxMap.get(word2);

        int i =0, j =0;
        int minDist = Integer.MAX_VALUE;

        while ( i<l1.size() && j<l2.size()){
            int a = l1.get(i);
            int b = l2.get(j);
            minDist = Math.min( minDist, Math.abs(a - b));
            if ( a > b )
                j++;
            else
                i++;
        }
        return minDist;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */